ui/src/pages/visualization.tsx

```diff
@@ -0,0 +1,175 @@
+import React, { useRef, useState, useEffect } from 'react';
+import ForceGraph3D from 'react-force-graph-3d';
+// import graphData from './output.json';
+import { Link } from 'react-router-dom';
+import { useUmap } from '../queries/useUmapData';
+import { useNavigate, useLocation } from 'react-router-dom';
+
+export interface GraphNode {
+  id: string;
+  name: string;
+  description: string;
+  cell_line: string;
+  assay?: string;
+  x: number;
+  y: number;
+  z: number;
+}
+
+export const Graph3D: React.FC = () => {
+  const { data: umap_data, isLoading, error } = useUmap();
+  const fgRef = useRef<any>();
+  const [searchId, setSearchId] = useState('');
+  const [message, setMessage] = useState('');
+  const navigate = useNavigate();
+  const location = useLocation();
+  const [colorByOption, setColorByOption] = useState<'cell_line' | 'assay'>('cell_line');
+
+
+  useEffect(() => {
+    const queryParams = new URLSearchParams(location.search);
+    const initialSearchId = queryParams.get('searchId');
+    if (initialSearchId) {
+      setSearchId(initialSearchId);
+      const timer = setTimeout(() => {
+        handleSearch();
+      }, 1000); // Give time for the graph to initialize
+      return () => clearTimeout(timer);
+    }
```

Comment:
Feels like a code-smell/anti-pattern. Any other way that we can do this?

Reply:
Theres gotta be a way for use to hook in to the graph and set some state once its initialized to avoid race conditions like this